crossdock/client/errorstchout/behavior.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package errorstchout
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/json"
+	"github.com/yarpc/yarpc-go/crossdock-go"
+	"github.com/yarpc/yarpc-go/crossdock/client/params"
+)
+
+const (
+	serverPort  = 8082
+	serviceName = "yarpc-test"
+)
+
+type test struct {
+	procedure string
+	body      []byte
+	validate  func(res3 []byte, isAppExpr bool, err error)
+}
+
+// Run exercises a YARPC server with outbound TChannel requests from a rigged
+// client and validates behavior that might only be visible to a TChannel
+// client without the YARPC abstraction interposed, typically errors.
+func Run(t crossdock.T) {
+	fatals := crossdock.Fatals(t)
+	assert := crossdock.Fatals(t)
+
+	tt := []test{
+		{
+			procedure: "echo",
+			body:      []byte{},
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.False(isAppErr, "malformed body must not be application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "malformed body must produce system error")
+				if !ok {
+					return
+				}
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(code, tchannel.ErrCodeBadRequest, "must produce bad request error")
+				assert.True(strings.Contains(err.Error(), `failed to decode "json"`))
+			},
+		},
+		{
+			procedure: "",
+			body:      []byte{},
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.False(isAppErr, "missing procedure must not produce an application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "missing procedure must produce system error")
+				if !ok {
+					return
+				}
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(code, tchannel.ErrCodeBadRequest, "missing procedure must produce bad request error")
```

Comment:
FYI, instead of prepending the error case name to all messages, you can add a name to all your test cases as an attribute and tag the `crossdock.T` with it when running the test.

```lang=go
t.Tag("name", tt.name)
```

That will be included in the output along with `status` and `output`.